api/proto/config/storage.proto

```diff
@@ -155,6 +155,12 @@ message ContentTree {
   // appear in the restAPI endpoint available at http://169.254.169.254. So user
   // application can read that data and treat the blob accordingly.
   string custom_meta_data = 10;
+
+  // List of DatastoreConfig UUIDs for fallback download if the download from
+  // the main DatastoreConfig pointed by the 'dsId' has been failed. Download
+  // from a datastore from the fallback list will be used one by one until
+  // success.
+  repeated string fallback_ds_ids_list = 11;
```

Comment:
@rouming IMO, it shouldn't be called fallback ds list, cloud will provide the list of ds only. For backward compatibility, it makes sense to fallback on the list if main ds field in empty. I believe logic should be if ds list is not empty then we'll not look at dsid field, if this list is empty then we'll look at dsid field (backward compatibility case). For now, cloud will provide both dsid and list of dsid but after the period of time dsid field will get deprecated.

Reply:
I decided to introduce a fallback list for two reasons:
1. From protocol perspective there is nothing to deprecate, no need to introduce any compatibility checks, although this is a very minor check, all such checks always look ugly and if there is a reason not to think about compatibility or introduce any checks or deprecate fields - better not to do so.
2. I want to emphasize on protocol level that there is a fallback datastore procedure, which is in place once the list is not empty. The fallback will be used only for specific cases, like air gap environment, and this is not a generic workflow. And all of this becomes clear once you see a "fallback" list and read a description.

Reply:
Please analyse from cloud perspective, image is an object of enterprise and if we're deciding to go with fallback mechanism then we have to make one datastore default for that particular image and it'll be true for all the devices on that enterprise which doesn't makes sense.

Reply:
Don't we introduce the fallback datastore specifically for air gap environment? The whole logic is about fallback: fail-fallback-retry.

Reply:
No, it'll be available across the board not only for air gapped environment.

Reply:
And the whole logic behind, is this not a fallback? We introduce an instance, with clear semantics of fallback (repeat with the next entry if fail). What is the other purpose of this enhancement rather than fallback? Regardless of how you manage it on cloud there will be always the datastore which will be tried first and then you fallback in case of failure.

Reply:
Let's put it straight: having the fallback semantics and implementing the fallback logic on eve I do not see why we should not call it other than fallback. Here I stick to clarity on protocol level.

Reply:
> Please analyse from cloud perspective, image is an object of enterprise and if we're deciding to go with fallback mechanism then we have to make one datastore default for that particular image and it'll be true for all the devices on that enterprise which doesn't makes sense.

Sorry, I cannot understand the point completely. In case of the particular image we should define datastore to pull the image from. After changes we will define the list of datastores for every particular image. So in all cases we are talking about configuration of particular image where we can (potentially) select defaults. Am I missing something?

Reply:
To me having a list of datastores is more general both at the UX and cloud side and on the EVE side. 
For the initial use case it is sufficient for EVE to try them in order until it finds a working one (basically it is an ordered list with an implied priority), but if we later have use cases where we want to optimize subsequent downloads we can make EVE remember which one worked last. Such an evolution is semantically harder if there is a primary+list of fallbacks; we'd have to explain to customers why it doesn't always try the primary first.

So a list of datastore ids makes sense to me; if this list is empty it implies an old controller and EVE/zedagent should use the deorecated single datastore ID to create the list used internally in EVE.

Reply:
We have only one particular use case in mind: fallback. The logic and semantics is all about the fallback. Optimization, ordering etc is orthogonal to that. Anyway, I created a separate PR with only protocol changes and with the 'ds_ids_list', so that 'dsId' becomes deprecated. 

@zed-rishabh 
> Please analyse from cloud perspective, image is an object of enterprise and if we're deciding to go with fallback > mechanism then we have to make one datastore default for that particular image 

Unfortunately you still __have__ to solve this problem for old devices, and for the sake of backwards compatibility set the 'dsId' field to the default one datastore.